pkg/joblib/library.go

```diff
@@ -0,0 +1,338 @@
+	"io"
+	"log/slog"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+)
+
+var (
+	// TODO: This should be loaded from some configuration and not hardcoded
+	logger = slog.New(slog.NewTextHandler(
+		os.Stdout,
+		&slog.HandlerOptions{
+			AddSource: false,
+			Level:     slog.LevelDebug,
+		},
+	))
+)
+
+const (
+	CgroupRoot = "/sys/fs/cgroup"
+
+	ExitSuccess      = 0
+	ExitProcessError = 1
+)
+
+// Job Contains information to interact with jobs.
+type Job struct {
+	id             string
+	processCmd     *exec.Cmd
+	forkCmd        *exec.Cmd
+	resourceLimits Resources
+	cgroup         *cgroup
+	stdout         io.Reader
+	stderr         io.Reader
+}
+
+// Resources cgroup limits that can be configured for jobs.
+type Resources struct {
+	MemoryBytes   uint64
+	CPUPercentage int
+	DiskIOBps     int
+}
+
+// cgroup Information used to construct a job's cgroup.
+type cgroup struct {
+	root       string
+	workerName string
+	jobID      string
+}
+
+// newCgroup Creates a new cgroup for the given job.
+func newCgroup(cgroupRoot string, workerName string, jobID string) (*cgroup, error) {
+	cg := &cgroup{
+		root:       cgroupRoot,
+		workerName: workerName,
+		jobID:      jobID,
+	}
+
+	jobPath := cg.withJobPath()
+
+	logger.Debug("Creating new cgroup", "path", jobPath)
+
+	if err := os.MkdirAll(jobPath, 0644); err != nil {
+		logger.Error("Failed to create cgroup", "err", err)
+
+		return nil, err
+	}
+
+	return cg, nil
+}
+
+// NewJob Create a new job to run the specified command using the given resource limits.
+func NewJob(workerName string, resourceLimits Resources, command string, args ...string) (*Job, error) {
+	logger.Debug("Creating new job", "workerName", workerName, "resourceLimits", resourceLimits, "command", command, "args", args)
+
+	id := uuid.NewString()
+	cg, err := newCgroup(CgroupRoot, workerName, id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Job{
+		id:             id,
+		processCmd:     exec.Command(command, args...),
+		forkCmd:        exec.Command("/proc/self/exe", "job"),
+		resourceLimits: resourceLimits,
+		cgroup:         cg,
+	}, nil
+}
+
+// ID Returns the ID of the job.
+func (j *Job) ID() string {
+	return j.id
+}
+
+// Start Begin execution of the job's command immediately.
+func (j *Job) Start() {
+	logger.Debug("Starting job", "args", os.Args, "ppid", os.Getppid(), "pid", os.Getpid())
+
+	if len(os.Args) > 1 && os.Args[1] == "job" {
+		logger.Info("Starting job process", "id", j.id, "process", j.processCmd.String())
+
+		j.runProcess()
+	} else {
+		logger.Info("Starting job", "id", j.id)
+
+		j.fork()
+	}
+}
+
+// Stop End execution of the job immediately.
+// FIXME: Handle the stop signal so that the job does not exit as a failure
+func (j *Job) Stop() error {
+	logger.Info("Stopping job", "id", j.id)
+
+	return j.forkCmd.Process.Signal(syscall.SIGTERM)
+}
+
+// Output Get the full output (stdout and stderr) from the job.
+func (j *Job) Output() (io.Reader, io.Reader) {
```

Comment:
This does not appear to support multiple simultaneous clients each streaming from the start of execution.

Reply:
Ah, sorry. I didn't put it together that supporting multiple concurrent clients was for the output. Is it acceptable to have these outputs written to disk?

Reply:
You can use as much memory or disk as you need. It's okay for the output of an individual job to grow unbounded, as long as the output is cleaned up when its no longer useful.

The API service does not need to scale, but we will evaluate the efficiency of the output streaming solution. I would think carefully about the ideal structure for solving this problem.